### PR TITLE
feat(resize): add ability to grow and shrink the tree when drawing

### LINKED ISSUE #865: Having option to grow window width automatically
Would be awesome if the window could resize itself as more directories get opened when there is not enough space.

I thought out auto_resize would accomplish this but that seems to have another purpose.

### PR DESCRIPTION
**Pull request by kyazdani42**
Will adapt the view size based on the longest line.
fixes #865

### CODE REVIEW

**Comment by kyazdani42** [comment]
found something, not sure if this is related, but with this option and the tree on the right side, if the window is too narrow, the live filter prompt is a bit messed up.

**Comment by kyazdani42** [comment]
maybe also take into account a min_width size ?

**Comment by alex-courtis** [comment]
Not related, but noticed a size change when using `view.width = 15`. The window will first display at width `17` (2 for sign column) but is then shrunk to width `15` when opening a file.

We could probably fix that in this PR.

[asciicast](https://asciinema.org/a/497739)

**Comment by alex-courtis** [comment]
> maybe also take into account a min_width size ?

That is simple and reliable.

We could be lazy and just use `view.width` as the minimum.

File: lua/nvim-tree/view.lua [hunk 1]
```diff
@@ -187,6 +188,23 @@ function M.open(options)
   events._dispatch_on_tree_open()
 end
 
+local function grow()
+  local lines = vim.api.nvim_buf_get_lines(M.get_bufnr(), 0, -1, false)
```
Comment (alex-courtis) [inline lua/nvim-tree/view.lua:192]: This includes the root folder. This results in a very wide window when you are in a deep path.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Tested OK in a variety of configurations e.g. git signs after.

**Comment by kyazdani42** [comment]
@alex-courtis i've fixed the naming and the initial width + the shrink issue.
This should now work as expected.
The issue on your screen recording looks weird to me, the width of the window is looking bugged to me, that might be a neovim issue (or maybe something because of the signcolumn ?).
I've noticed that neovim takes the signcolumn into account when setting the width.

**Comment by alex-courtis** [comment]
> The issue on your screen recording looks weird to me, the width of the window is looking bugged to me, that might be a neovim issue (or maybe something because of the signcolumn ?).

It only happens for very narrow windows. It's not worth looking at now; just thought I'd record my findings.

**Review by alex-courtis — COMMENTED** [review]
Shrinking is not reliable / happens at the wrong time.

Defaults with `adaptive_size`:
1293-length-test

Extract [1293-length-test.tar.gz](https://github.com/kyazdani42/nvim-tree.lua/files/8792626/1293-length-test.tar.gz) to /tmp to test.

**Comment by kyazdani42** [comment]
Shrinking should be reliable now !

File: lua/nvim-tree/view.lua [hunk 2]
```diff
@@ -373,8 +392,10 @@ end
 
 function M.setup(opts)
   local options = opts.view or {}
+  M.View.adaptive_size = options.adaptive_size
   M.View.side = options.side
   M.View.width = options.width
+  M.View.initial_width = options.width
```
Comment (alex-courtis) [inline lua/nvim-tree/view.lua:398]: `view.width = "20%"` is a problem.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Working nicely. I look forward to using this.

File: lua/nvim-tree/view.lua (hunk 2 above)
Comment (kyazdani42) [reply lua/nvim-tree/view.lua:398]: indeed ! Completely forgot about this one

**Comment by kyazdani42** [comment]
update initial width to use `get_size`

**Review by alex-courtis — APPROVED** [review]
👍

**Pull request merged**